### PR TITLE
Advanced functions - exercise 1

### PR DESCRIPTION
**Pull request by SaraMojsova**

### CODE REVIEW

File: js-advanced-functions/fn-exercise-1.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    //     // console.log(el)
+    // }
+    // the two for cycles do the same
+    newArray=array
+    console.log('Incrementing function by 2',newArray)
+}
+incrementing(initialData,padding=2)
+
+let newArray2=[]
+initialData=[200,500,300,1000]
```
Comment (atodoroski) [inline js-advanced-functions/fn-exercise-1.js:23]:
```suggestion
```

File: js-advanced-functions/fn-exercise-1.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
+//1. part one
+let initialData=[200,500,300,1000]
+console.log('INITIAL DATA: ',initialData)
+let newArray=[]
```
Comment (atodoroski) [inline js-advanced-functions/fn-exercise-1.js:4]:
```suggestion
```

File: js-advanced-functions/fn-exercise-1.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    }
+    // for(el of array){
+    //     el=el+padding
+    //     // console.log(el)
+    // }
+    // the two for cycles do the same
+    newArray=array
+    console.log('Incrementing function by 2',newArray)
+}
+incrementing(initialData,padding=2)
```
Comment (atodoroski) [inline js-advanced-functions/fn-exercise-1.js:20]:
```suggestion
incrementing(initialData, 2)
```

File: js-advanced-functions/fn-exercise-1.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+let newArray2=[]
+initialData=[200,500,300,1000]
+
+const decrementing=(array,padding)=>{
```
Comment (atodoroski) [inline js-advanced-functions/fn-exercise-1.js:25]: You can apply the same changes as for the incrementing function

File: js-advanced-functions/fn-exercise-1.js (hunk 3 above)
Comment (atodoroski) [reply js-advanced-functions/fn-exercise-1.js:20]: When calling a function the passing of the arguments is by passing a variable or an inline value. What was wrong here is the `padding = 2`

File: js-advanced-functions/fn-exercise-1.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    newArray2=array
+    console.log('Decrementing function by 2',newArray2)
+}
+decrementing(initialData,padding=2)
+
+//2.part two 
+//HOF
+console.log('Part 2 of the exercise --------------')
+let newArray3=[]
+initialData=[200,500,300,1000]
```
Comment (atodoroski) [inline js-advanced-functions/fn-exercise-1.js:39]:
```suggestion
```

File: js-advanced-functions/fn-exercise-1.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    }
+    newArray2=array
+    console.log('Decrementing function by 2',newArray2)
+}
+decrementing(initialData,padding=2)
+
+//2.part two 
+//HOF
+console.log('Part 2 of the exercise --------------')
+let newArray3=[]
```
Comment (atodoroski) [inline js-advanced-functions/fn-exercise-1.js:38]:
```suggestion
```

File: js-advanced-functions/fn-exercise-1.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+initialData=[200,500,300,1000]
+
+const operationInc = (array,padding)=>{
```
Comment (atodoroski) [inline js-advanced-functions/fn-exercise-1.js:41]: You need to apply the following refactor to this function. 
The incrementing function should receive two parameters, one for a number and the other for the padding.
It should return a value `num + padding`. There is no iterration here. The iterration will be done in the main function

File: js-advanced-functions/fn-exercise-1.js [hunk 8]
```diff
@@ -0,0 +1,65 @@
+//1. part one
+let initialData=[200,500,300,1000]
+console.log('INITIAL DATA: ',initialData)
+let newArray=[]
+console.log('Part 1 of the exercise -------------')
+
+const incrementing=(array,padding)=>{
+    for(i=0;i<array.length;i++){
+        array[i]=array[i]+padding
```
Comment (atodoroski) [inline js-advanced-functions/fn-exercise-1.js:9]:
```suggestion
        newArray.push(array[i] + padding);
```

File: js-advanced-functions/fn-exercise-1.js [hunk 9]
```diff
@@ -0,0 +1,65 @@
+//1. part one
+let initialData=[200,500,300,1000]
+console.log('INITIAL DATA: ',initialData)
+let newArray=[]
+console.log('Part 1 of the exercise -------------')
+
+const incrementing=(array,padding)=>{
+    for(i=0;i<array.length;i++){
```
Comment (atodoroski) [inline js-advanced-functions/fn-exercise-1.js:8]:
```suggestion
    const newArray = [];
    for(i=0;i<array.length;i++){
```

File: js-advanced-functions/fn-exercise-1.js [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+        array[i]=array[i]+padding
+    }
+    newArray3=array
+    return newArray3    
+}
+// 
+let newArray4=[]
+initialData=[200,500,300,1000]
+
+const operationDec = (array,padding)=>{
```
Comment (atodoroski) [inline js-advanced-functions/fn-exercise-1.js:52]: You need to apply the same refactor as for the increment

File: js-advanced-functions/fn-exercise-1.js [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    for(i=0;i<array.length;i++){
+        array[i]=array[i]-padding;
+        // console.log('initial data',initialData)
+    }
+    newArray4=array
+    return newArray4    
+}//is not working
+const main =(array,fn,padding)=>{
```
Comment (atodoroski) [inline js-advanced-functions/fn-exercise-1.js:60]: Here you need to iterrate over the array and call the proper operation (fn parameter) to apply the correct function. 
The result should be printed in new array

Comment (SaraMojsova) [reply js-advanced-functions/fn-exercise-1.js:60]: @atodoroski  This part is not really clear to me

File: js-advanced-functions/fn-exercise-1.js (hunk 7 above)
Comment (SaraMojsova) [reply js-advanced-functions/fn-exercise-1.js:41]: @atodoroski Also here, should the code be: 
const operationInc = (number,padding)=>{
  // not sure what to insert here
    return number+padding   
}

Comment (atodoroski) [reply js-advanced-functions/fn-exercise-1.js]: nothing, the function has just one line. Then you will call this function to apply the increment/decrement operation

File: js-advanced-functions/fn-exercise-1.js [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+    for(const elem of array){
+        newArray.push(fn(elem,padding))
+    }
+    return newArray
+}
+const incrHOF=main(initialData,operationInc,3)
+const decrHOF=main(initialData,operationDec,3)
+console.log(incrHOF)
+console.log(decrHOF)
+// Where to print the message `HOF call (increment) by 3` and `HOF call (increment) by 3` ?
```
Comment (atodoroski) [inline js-advanced-functions/fn-exercise-1.js:53]: You did everything well, just the final part.
`const main = (array, fn, padding) => {
    const newArray = []
    for (const elem of array) {
        newArray.push(fn(elem, padding))
    }
    return (fn) => { // here you return an array but you need to return new function
        return fn(newArray); // this function will be our sort function
    }
}
const incrHOF = main(initialData, operationInc, 3)
const decrHOF = main(initialData, operationDec, 3)
console.log(incrHOF(sort));
console.log(decrHOF(sort))`
What is left here is to create a sort function and use it in the console logs above. This is the way how HOF are working

**Review by atodoroski — APPROVED** [review]
Now it's perfect. Thank you!

**Pull request merged**